CDAP-12109 Updates to the SmartWorkflow to include condition phases.

sagarkapare opened the pull request.

sagarkapare committed:
CDAP-12109 Updates to the SmartWorkflow to include condition phases.

sagarkapare commented:
Pending test cases, will add them as another commit. @albertshau please take a look when you get a chance. thanks!

albertshau commented on cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java:
```diff
@@ -413,19 +404,33 @@ private void addProgram(String phaseName, WorkflowProgramAdder programAdder) {
       programAdder.addAction(new PipelineAction(batchPhaseSpec));
+    } else if (pluginTypes.contains(Condition.PLUGIN_TYPE)) {
+      // conditions will be all by themselves in a phase
+      addCondition(programAdder, programName, batchPhaseSpec);
```
seems like you should be passing in phaseName

sagarkapare replied:
fixed.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/condition/PipelineCondition.java:
```diff
@@ -16,17 +16,93 @@
+      }
+      return result;
+    } catch (Exception e) {
+      String msg = String.format("Error executing condition '%s' in the pipeline.", phaseSpec.getPhaseName());
```
stageSpec.getName()

sagarkapare replied:
fixed.

sagarkapare commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/condition/PipelineCondition.java:
```diff
@@ -16,17 +16,93 @@
+      // should not happen
+      throw new IllegalStateException("WorkflowContext for the Condition cannot be null.");
+    }
+    Map<String, String> properties = input.getConditionSpecification().getProperties();
```
I realized that we need an api change for the WorkflowContext to get the ConditionSpecifications. Have added it in this PR.

sagarkapare commented:
@albertshau Addressed comment and added test cases. please take a look again. thanks!

albertshau commented on cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java:
```diff
@@ -413,19 +416,33 @@ private void addProgram(String phaseName, WorkflowProgramAdder programAdder) {
       programAdder.addAction(new PipelineAction(batchPhaseSpec));
+    } else if (pluginTypes.contains(Condition.PLUGIN_TYPE)) {
+      // conditions will be all by themselves in a phase
+      addCondition(programAdder, phaseName, batchPhaseSpec);
```
I think its cleaner if you just do programAdder.condition() here, and move the other logic in addCondition() to the addPrograms() method. That way all the logic related to forking or condition-ing is in the same place

sagarkapare replied:
sounds good.. will update it

sagarkapare replied:
updated the code.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/condition/PipelineCondition.java:
```diff
@@ -16,17 +16,93 @@
+      if (token == null) {
+        throw new IllegalStateException("WorkflowToken cannot be null when Condition is executed through Workflow.");
+      }
+      for (Map.Entry<String, String> entry : arguments.getAddedArguments().entrySet()) {
```
don't think this is needed, as ConditionContext can only get Arguments but not SettableArguments

sagarkapare replied:
Updated `ConditionContext` to return `SettableArguments` since seems useful to have it.

albertshau commented on cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/condition/MockCondition.java:
```diff
@@ -0,0 +1,71 @@
+    private String name;
+  }
+
+  public static final PluginClass PLUGIN_CLASS = getPluginClass();
```
nit: for consistency, move this to the top of the class

albertshau commented on cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java:
```diff
@@ -619,6 +621,116 @@ private void testPipelineWithActions(Engine engine) throws Exception {
+    DataSetManager<Table> sinkManager = getDataset("sink3Output");
+    Set<StructuredRecord> expected = ImmutableSet.of(recordSamuel, recordBob);
+    Set<StructuredRecord> actual = Sets.newHashSet(MockSink.readOutput(sinkManager));
+    Assert.assertEquals(expected, actual);
```
should also check that the other sinks have no data

sagarkapare replied:
added checks for other sinks to be empty.

sagarkapare commented:
@albertshau Addressed comments. please take a look again. thanks!

albertshau reviewed (approved):
lgtm

sagarkapare commented:
Rebased against release/4.3 branch and squashed commits. New build is running here -https://builds.cask.co/browse/CDAP-RUT1247-2. Will merge once the build succeeds.

sagarkapare commented:
Build failed. looking into it.

sagarkapare commented:
Same dataset was being used in multiple tests causing incorrect count in the metrics. Fixed and new build is running here - https://builds.cask.co/browse/CDAP-RUT1247-3

sagarkapare commented:
Build is green https://builds.cask.co/browse/CDAP-RUT1247-4. Squashing and merging.
